src/cmd/tools/doctest/doctest.go

```diff
@@ -0,0 +1,142 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+)
+
+const (
+	operationRegEx  = `(?s)<operation>.*?</operation>|(?s)<operation_json>.*?</operation_json>`
+	validationRegEx = `(?s)<validation>.*?</validation>`
+
+	operationOpen      = `<operation>`
+	operationClose     = `</operation>`
+	operationJSONOpen  = `<operation_json>`
+	operationJSONClose = `</operation_json>`
+	validationOpen     = `<validation>`
+	validationClose    = `</validation>`
+
+	all = `<operation>|</operation>|<validation>|</validation>`
+)
+
+var (
+	file string
+)
+
+func init() {
+	flag.StringVar(&file, "file", "", "File to be parsed")
+}
+
+func main() {
+	flag.Parse()
+
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Fatalf("unable to read source file: %v\n", err)
+	}
+
+	validationRegEx, err := bashScript(f, file)
+	if err != nil {
+		log.Fatalf("unable to create bash script: %v", err)
+	}
+
+	if err := markdownFile(f, file, validationRegEx); err != nil {
+		log.Fatalf("unable to create markdown file: %v", err)
+	}
+}
+
+func markdownFile(contents []byte, fileName string, validationRegEx *regexp.Regexp) error {
+	// create new markdown file based on name of source file
+	out, err := os.Create(strings.Replace(fileName, ".source", "", -1))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Chmod(0644)
+
+	// convert tags to markdown syntax
+	OpOpen := regexp.MustCompile(operationOpen)
+	OpClose := regexp.MustCompile(operationClose)
+	OpJSONOpen := regexp.MustCompile(operationJSONOpen)
+	OpJSONClose := regexp.MustCompile(operationJSONClose)
+
+	contents = OpOpen.ReplaceAll(contents, []byte("```"))
+	contents = OpClose.ReplaceAll(contents, []byte("```"))
+	contents = OpJSONOpen.ReplaceAll(contents, []byte("```json"))
+	contents = OpJSONClose.ReplaceAll(contents, []byte("```"))
+	contents = validationRegEx.ReplaceAll(contents, nil)
+	out.Write(contents)
+
+	return nil
+}
+
+func bashScript(contents []byte, fileName string) (*regexp.Regexp, error) {
+	// create bash script file based on name of source file
+	scriptFileName := fmt.Sprintf("%s%s", strings.Replace(fileName, ".md.source", "", -1), ".sh")
+	script, err := os.Create(scriptFileName)
+	if err != nil {
+		return nil, err
+	}
+	defer script.Close()
+
+	script.Chmod(0644)
+	script.WriteString("#!/usr/bin/env bash\n\n")
```

Comment:
take a file which is like "script header template" rather than embed this into go code